corehq/warehouse/models/facts.py

```diff
@@ -122,34 +116,30 @@ class ApplicationStatusFact(BaseFact, CustomSQLETLMixin):
 
     Grain: app_id, user_id
     '''
-    # TODO: Write Update SQL Query (currently there exists a placeholder)
     slug = APP_STATUS_FACT_SLUG
 
-    # TODO: Add app dimension
-    # app_dim = models.CharField(max_length=255)
+    app_id = models.CharField(max_length=255, null=True, db_index=True)
 
-    # TODO: Add domain dimension
-    # domain_dim = models.CharField(max_length=255)
+    domain = models.CharField(max_length=255, null=True, db_index=True)
 
-    user_dim = models.ForeignKey(UserDim, on_delete=models.PROTECT)
+    user_id = models.CharField(max_length=255, null=True, db_index=True)
```

Comment:
it looks like user_id and app_id are still denormalized - just curious why that was done? if we ever want to join to the user dim or app dim it will take much longer than if we joined on integer ids

Reply:
would it work to index those fields on the dim table, for faster joins. I left it like this because I think the vast majority of queries, if not every one, will filter the app status report on user or app or both, and our existing filters output app_id and user_id for those, so each report view would require either a second query to the dim to get the primary key or a join. this schema allows us to do a single query with no joins and get the info we want. I think the alternative is to rewrite the hq filters for reports that use the warehouse or take the performance hit and do the join. the first seems out of scope at this point, and i was not sure if the second was worth it.

Reply:
ok if that's the only use case it's probably fine, but only because this is for a very specialized report and probably won't be used anywhere else. if this were a more reusable fact table it would be more important that all foreign keys were integer ids. i also wouldn't recommend joining this to the user dim or app dim with these ids, if that's ever needed then the dim foreign keys should be used and the join would have to happen before filtering.

Reply:
it turns out I needed more information than expected from the user dim, so i am in fact going back to make these normal foreign keys. sorry about all the misdirection